### PR TITLE
Message types

### PR DESCRIPTION
**Pull request by szarnyasg**

### CODE REVIEW

File: ire/src/main/scala/hu/bme/mit/ire/nodes/unary/MapperNode.scala [hunk 1]
```diff
@@ -14,13 +14,19 @@ import hu.bme.mit.ire.messages.{ChangeSet, ReteMessage}
...
-                ) extends UnaryNode with SingleForwarder {
+                ) extends UnaryNode[DataMessage] with SingleForwarder {
   override def onSizeRequest() = 0
 
-  def onChangeSet(changeSet: ChangeSet): Unit = {
-    forward(ChangeSet(
-      changeSet.positive.map(t => t.updated(index, function(t(index)))),
-      changeSet.negative.map(t => t.updated(index, function(t(index))))
+  def onChangeSet(changeSet: DataMessage): Unit = {
+//    forward(ChangeSet(
```
Comment (wafle) [inline ire/src/main/scala/hu/bme/mit/ire/nodes/unary/MapperNode.scala]: Please remove these.

File: ire/src/main/scala/hu/bme/mit/ire/messages/Messages.scala [hunk 2]
```diff
@@ -9,7 +9,23 @@ import scala.collection.mutable
 
 class ReteMessage() {}
 
-case class ChangeSet(positive: TupleBag = Vector(), negative: TupleBag = Vector()) extends ReteMessage
+trait DataMessage extends ReteMessage {
+  def changeSets: Seq[TupleBag]
+  def createNew(bags: Seq[TupleBag]): DataMessage
```
Comment (wafle) [inline ire/src/main/scala/hu/bme/mit/ire/messages/Messages.scala]: Let's write a map function that forwards the parameter function to all the bags.

File: ire/src/main/scala/hu/bme/mit/ire/nodes/binary/JoinNode.scala [hunk 3]
```diff
@@ -14,17 +14,17 @@ class JoinNode(override val next: (ReteMessage) => Unit,
               ) extends JoinNodeBase with SingleForwarder {
 }
 
-class ParallelJoinNode(override val children: Vector[(ReteMessage) => Unit],
```
Comment (wafle) [inline ire/src/main/scala/hu/bme/mit/ire/nodes/binary/JoinNode.scala]: Why is this removed? It's used in JoinNodeTest, why is this not breaking?

File: ire/src/main/scala/hu/bme/mit/ire/nodes/unary/SelectionNode.scala [hunk 4]
```diff
@@ -2,17 +2,19 @@ package hu.bme.mit.ire.nodes.unary
...
-                   ) extends UnaryNode with SingleForwarder {
+                   ) extends UnaryNode[DataMessage] with SingleForwarder {
   override def onSizeRequest() = 0
-  def onChangeSet(changeSet: ChangeSet): Unit = {
-    forward(ChangeSet(
-      changeSet.positive.filter(condition),
-      changeSet.negative.filter(condition)
-    ))
+  def onChangeSet(changeSet: DataMessage): Unit = {
+//    forward(ChangeSet(
```
Comment (wafle) [inline ire/src/main/scala/hu/bme/mit/ire/nodes/unary/SelectionNode.scala]: Remove these.

File: ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala [hunk 5]
```diff
@@ -1,14 +1,14 @@
 package hu.bme.mit.ire.stateless.unary
 
 import hu.bme.mit.ire.SingleForwarder
-import hu.bme.mit.ire.messages.{ChangeSet, ReteMessage}
+import hu.bme.mit.ire.messages.{BatchChangeSet, ReteMessage}
 import hu.bme.mit.ire.nodes.unary.UnaryNode
 
-class DuplicateEliminationNode(override val next: (ReteMessage) => Unit) extends UnaryNode with SingleForwarder {
+class DuplicateEliminationNode(override val next: (ReteMessage) => Unit) extends UnaryNode[BatchChangeSet] with SingleForwarder {
```
Comment (wafle) [inline ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala]: Negative messages still go through this.

File: ire/src/test/scala/hu/bme/mit/ire/nodes/binary/JoinNodeTest.scala [hunk 6]
```diff
@@ -119,26 +119,26 @@ class JoinNodeTest(_system: ActorSystem) extends TestKit(_system) with ImplicitS
...
-
-      forkingJoiner ! Primary(ChangeSet(positive = tupleBag(tuple(0, 2))))
-      forkingJoiner ! Primary(ChangeSet(positive = tupleBag(tuple(1, 2))))
-
-      joinerB ! Secondary(ChangeSet(positive = tupleBag(tuple(1, 3))))
-      expectMsg(ChangeSet(positive = tupleBag(tuple(1, 2, 3, 3))))
-      joinerA ! Secondary(ChangeSet(positive = tupleBag(tuple(0, 3))))
-      expectMsg(ChangeSet(positive = tupleBag(tuple(0, 2, 2, 3))))
-    }
+//    "do parallel joins" in {
```
Comment (wafle) [inline ire/src/test/scala/hu/bme/mit/ire/nodes/binary/JoinNodeTest.scala]: Oh i see what's happening to it. Please don't merge until this is fixed.

File: .travis.yml [hunk 7]
```diff
@@ -22,7 +22,7 @@ install:
   - ./gradlew assemble testClasses --stacktrace
 script:
   # only build and deploy the results if the build is performed on the master branch
-#  - if scripts/travis-tex-run-required.sh ; then scripts/generate-and-deploy-report.sh; fi
+  - if scripts/travis-tex-run-required.sh ; then scripts/generate-and-deploy-report.sh; fi
```
Comment (wafle) [inline .travis.yml]: Does this report work? It's not tested by this PR, right?

File: scripts/generate-and-deploy-report.sh [hunk 8]
```diff
@@ -6,10 +6,10 @@ set -e
 
 pushd .
 
-./gradlew report --stacktrace --quiet
-cd opencypher-report
-make texfot
-cd ..
+#./gradlew report --stacktrace --quiet
+#cd opencypher-report
+#make texfot
```
Comment (wafle) [inline scripts/generate-and-deploy-report.sh]: Are we ever going to need this? We can still restore it from git history.

File: .travis.yml (hunk 7 above)
Comment (szarnyasg) [reply .travis.yml]: The point if this is to deploy the homepage to http://docs.inf.mit.bme.hu/ingraph/. The tech report will not work anytime soon, so I'll only invoke deploy.sh.

File: ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala (hunk 5 above)
Comment (szarnyasg) [reply ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala]: What do you mean?

Comment (wafle) [reply ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala]: If this node is in a network of ire nodes, it might receive IncerementalChangeSets with negative elements.

**Comment by szarnyasg** [comment]
> I don't get why this is required. To avoid a Vector() allocation on every message? I would expect that to have a negligeble effect.

No, the reason for this PR is not performance but issue #212.

There are stateless nodes that can be used for both incremental and non-incremental processing (such as Projection node and Selection node), so these should be able to produce both Batch and Incremental changesets. In constrast, some will only work for either Batch or Incremental.

File: ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala (hunk 5 above)
Comment (szarnyasg) [reply ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala]: Let's suppose the EngineFactory takes care of that. This would be quite cumbersome to express in the type system alone.

Comment (wafle) [reply ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala]: Let's not. Why would we do that?

>quite cumbersome to express in the type system alone

Well it was expressed before this change.

Comment (szarnyasg) [reply ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala]: Any ideas? Perhaps we should introduce a trait that applies to "control messages" and "batch change sets"?

Comment (wafle) [reply ire/src/main/scala/hu/bme/mit/ire/stateless/unary/DuplicateEliminationNode.scala]: I don't have any ideas because i still don't get what this PR is trying to achieve.

**Comment by szarnyasg** [comment]
It's a preparation for https://github.com/FTSRG/ingraph/issues/212#issuecomment-328292369.

For non-incremental processing, we should only use change sets that have a single positive change set. For incremental, we should stick to good old positive-negative change sets.

**Comment by wafle** [comment]
Here you go:

```scala
trait ChangeSet extends ReteMessage {
  val positive: TupleBag
  val negative: TupleBag
}
case class IncrementalChange(positive: TupleBag = Vector(), negative: TupleBag = Vector()) extends ChangeSet

case class NonIcrementalChange(changes: TupleBag = Vector()) extends ChangeSet {
  override val negative = Vector()
}
```

**Comment by szarnyasg** [comment]
I like the pragmatism of this approach. But in that case, is there a way to enforce that incremental nodes (such as join) can only operate on `IncrementalChange`s? Is there even a need to enforce this?

**Comment by wafle** [comment]
We could `assert`... 🙃 
Actually we should probably implement the map function in `ChangeSet`. All stateless node are symmetric on positive/negative, and if we use the mapper in those, we can just raise an exception if a NonIcremental negative is accessed.

**Comment by szarnyasg** [comment]
> All stateless node are symmetric on positive/negative

Precisely, that's why I introduced the `bag` function @ https://github.com/FTSRG/ingraph/blob/message-types/ire/src/main/scala/hu/bme/mit/ire/messages/Messages.scala#L13

**Comment by wafle** [comment]
Yeah. I'm saying we should combine that with createNew to form a map function.

**Comment by szarnyasg** [comment]
Closing this for now, we'll first refactor the engine and try this again later.
